File: TileCalorimeter/TileMonitoring/python/TileDigitsFlxMonitorAlgorithm.py
#
#  Copyright (C) 2002-2024 CERN for the benefit of the ATLAS collaboration
#

'''
@file TileDigitsFlxMonitorAlgorithm.py
@brief Python configuration of TileDigitsFlxMonitorAlgorithm algorithm for the Run III
'''

def TileDigitsFlxMonitoringConfig(flags, fragIDs=[0x201, 0x402], **kwargs):
    '''Function to configures TileDigitsFlxMonAlg algorithms in the monitoring system.'''
    
    kwargs.setdefault('TileDigitsContainerLegacy', 'TileDigitsCnt')
    kwargs.setdefault('TileDigitsContainerFlx', 'TileDigitsFlxCnt')
    kwargs.setdefault('FirstSample', 0)
    kwargs.setdefault('LastSample', 15)
    kwargs.setdefault('FelixOffset', 0)
    kwargs.setdefault('FelixScale', 4)

    felixScale = kwargs['FelixScale']

    from AthenaConfiguration.ComponentAccumulator import ComponentAccumulator
    result = ComponentAccumulator()

    from TileConditions.TileCablingSvcConfig import TileCablingSvcCfg
    result.merge( TileCablingSvcCfg(flags) )

    from AthenaMonitoring import AthMonitorCfgHelper
    helper = AthMonitorCfgHelper(flags, 'TileDigitsFlxMonitoring')

    from AthenaConfiguration.ComponentFactory import CompFactory
    TileDigitsFlxMonitorAlgorithm = CompFactory.TileDigitsFlxMonitorAlgorithm
    tileDigitsFlxMonAlg = helper.addAlgorithm(TileDigitsFlxMonitorAlgorithm, 'TileDigitsFlxMonAlg')
    tileDigitsFlxMonAlg.TriggerChain = ''

    for k, v in kwargs.items():
        setattr(tileDigitsFlxMonAlg, k, v)

    # Configure histogram with TileDigitsFlxMonAlg algorithm execution time
    executeTimeGroup = helper.addGroup(tileDigitsFlxMonAlg, 'TileDigitsFlxMonExecuteTime', 'Tile/Felix')
    executeTimeGroup.defineHistogram('TIME_execute', path = 'Digits', type='TH1F',
                                     title = 'Time for execute TileDigitsFlxMonAlg algorithm;time [#mus]',
                                     xbins = 300, xmin = 0, xmax = 300000)

    runNumber = flags.Input.RunNumbers[0]
    from TileCalibBlobObjs.Classes import TileCalibUtils as Tile

    modules = []
    if fragIDs:
        for fragID in fragIDs:
            ros = fragID >> 8
            drawer = fragID & 0x3F
            modules += [Tile.getDrawerString(ros, drawer)]
    else:
        for ros in range(1, Tile.MAX_ROS):
            for drawer in range(0, Tile.MAX_DRAWER):
                modules += [Tile.getDrawerString(ros, drawer)]

    channelHFNGroup = helper.addGroup(tileDigitsFlxMonAlg, 'TileFlxMonHFN', 'Tile/Felix/Digits')
    for moduleName in modules:
        for gainName in ['HG', 'LG']:
            title = f'Run {runNumber} {moduleName} {gainName}: Mean RMS in event (HFN)-FELIX;Channel;HFN'
            name = f'{moduleName}_{gainName}_channel,{moduleName}_{gainName}_HFN;{moduleName}_HFN_{gainName}'
            path = moduleName
            channelHFNGroup.defineHistogram(name, title = title, path = path, type = 'TProfile',
                                            xbins = 48, xmin = -0.5, xmax = 47.5)

    channelPedGroup = helper.addGroup(tileDigitsFlxMonAlg, 'TileFlxMonPed', 'Tile/Felix/Digits')
    for moduleName in modules:
        for gainName in ['HG', 'LG']:
            title = f'Run {runNumber} {moduleName} {gainName}: Pedestal, sample[0]-FELIX;Channel;ADC'
            name = f'{moduleName}_{gainName}_channel,{moduleName}_{gainName}_Pedestal;{moduleName}_Pedestal_{gainName}'
            path = moduleName
            channelPedGroup.defineHistogram(name, title = title, path = path, type = 'TProfile',
                                            xbins = 48, xmin = -0.5, xmax = 47.5)

    channelHFNGroup = helper.addGroup(tileDigitsFlxMonAlg, 'TileLegacyMonHFN', 'Tile/Legacy/Digits')
    for moduleName in modules:
        for gainName in ['HG', 'LG']:
            title = f'Run {runNumber} {moduleName} {gainName}: Mean RMS in event (HFN);Channel;HFN'
            name = f'{moduleName}_{gainName}_channel,{moduleName}_{gainName}_HFN;{moduleName}_HFN_{gainName}'
            path = moduleName
            channelHFNGroup.defineHistogram(name, title = title, path = path, type = 'TProfile',
                                            xbins = 48, xmin = -0.5, xmax = 47.5)

    channelPedGroup = helper.addGroup(tileDigitsFlxMonAlg, 'TileLegacyMonPed', 'Tile/Legacy/Digits')
    for moduleName in modules:
        for gainName in ['HG', 'LG']:
            title = f'Run {runNumber} {moduleName} {gainName}: Pedestal, sample[0];Channel;ADC'
            name = f'{moduleName}_{gainName}_channel,{moduleName}_{gainName}_Pedestal;{moduleName}_Pedestal_{gainName}'
            path = moduleName
            channelPedGroup.defineHistogram(name, title = title, path = path, type = 'TProfile',
                                            xbins = 48, xmin = -0.5, xmax = 47.5)

    channelSamplesFlxGroup = helper.addGroup(tileDigitsFlxMonAlg, 'TileFlxMonSamples', 'Tile/Felix/Digits')
    for moduleName in modules:
        for channel in range(0, Tile.MAX_CHAN):
            for gainName in ['HG', 'LG']:
                title = (f'Run {runNumber} {moduleName} channel {channel} {gainName}: Samples-FELIX;Sample [ADC];N')
                name = f'{moduleName}_ch_{str(channel)}_{gainName}_samples'
                path = moduleName
                channelSamplesFlxGroup.defineHistogram(name, title = title, path = path, type = 'TH1F',
                                                       xbins = 4096, xmin = -0.5, xmax = 4095.5)


    channelSamplesLegacyGroup = helper.addGroup(tileDigitsFlxMonAlg, 'TileLegacyMonSamples', 'Tile/Legacy/Digits')
    for moduleName in modules:
        for channel in range(0, Tile.MAX_CHAN):
            for gainName in ['HG', 'LG']:
                title = (f'Run {runNumber} {moduleName} channel {channel} {gainName}: Samples;Sample [ADC];N')
                name = f'{moduleName}_ch_{str(channel)}_{gainName}_samples'
                path = moduleName
                channelSamplesLegacyGroup.defineHistogram(name, title = title, path = path, type = 'TH1F',
                                                          xbins = 1024, xmin = -0.5, xmax = 1023.5)

    channelAllSamplesDiffGroup = helper.addGroup(tileDigitsFlxMonAlg, 'TileChannelAllSamplesDiff', 'Tile/Compare/Digits')
    for moduleName in modules:
        for channel in range(0, Tile.MAX_CHAN):
            for gainName in ['HG', 'LG']:
                title = (f'Run {runNumber} {moduleName} channel {channel} {gainName}: Samples difference (FELIX-Legacy*{felixScale});Sample [ADC];N')
                name = f'{moduleName}_ch_{str(channel)}_{gainName}_samples_diff'
                path = moduleName
                channelAllSamplesDiffGroup.defineHistogram(name, title = title, path = path, type = 'TH1F',
                                                           xbins = 200, xmin = -100, xmax = 100)

    moduleAllSamplesGroup = helper.addGroup(tileDigitsFlxMonAlg, 'TileModuleAllSamplesDiff', 'Tile/Compare/Digits')
    for moduleName in modules:
        for gainName in ['HG', 'LG']:
            title = (f'Run {runNumber} {moduleName} {gainName}: All channels samples difference (FELIX-Legacy*{felixScale});Sample [ADC];N')
            name = f'{moduleName}_{gainName}_samples_diff;{moduleName}_{gainName}_all_channels_samples_diff'
            path = moduleName
            moduleAllSamplesGroup.defineHistogram(name, title = title, path = path, type = 'TH1F',
                                                  xbins = 200, xmin = -100, xmax = 100)

    moduleSamplesDiffGroup = helper.addGroup(tileDigitsFlxMonAlg, 'TileModuleSamplesDiff', 'Tile/Compare/Digits')
    for moduleName in modules:
        for gainName in ['HG', 'LG']:
            title = f'Run {runNumber} {moduleName} {gainName}: Samples difference (FELIX-Legacy*{felixScale});Channel;ADC'
            name = f'{moduleName}_{gainName}_channel,{moduleName}_{gainName}_samples_diff;{moduleName}_{gainName}_samples_diff'
            path = moduleName
            moduleSamplesDiffGroup.defineHistogram(name, title = title, path = path, type = 'TProfile',
                                                   xbins = 48, xmin = -0.5, xmax = 47.5)


    accumalator = helper.result()
    result.merge(accumalator)
    return result


if __name__=='__main__':    
    # Setup logs
    from AthenaCommon.Logging import log
    from AthenaCommon.Constants import INFO
    log.setLevel(INFO)

    # Set the Athena configuration flags
    from AthenaConfiguration.AllConfigFlags import initConfigFlags
    flags = initConfigFlags()
    parser = flags.getArgumentParser()
    parser.add_argument('--postExec', help='Code to execute after setup')
    parser.add_argument('--digits', default="TileDigitsCnt", help='Tile digits container to be monitored')
    parser.add_argument('--frag-ids', dest='fragIDs', nargs="*", default=['0x201','0x402'], help='Tile Frag IDs of modules to be monitored. Empty=ALL')
    args, _ = parser.parse_known_args()

    fragIDs = [int(fragID, base=16) for fragID in args.fragIDs]

    from AthenaConfiguration.TestDefaults import defaultGeometryTags, defaultTestFiles
    flags.Input.Files = defaultTestFiles.RAW_RUN2
    flags.GeoModel.AtlasVersion = defaultGeometryTags.RUN2
    flags.Output.HISTFileName = 'TileDigitsFlxMonitorOutput.root'
    flags.DQ.useTrigger = False
    flags.DQ.enableLumiAccess = False
    flags.Exec.MaxEvents = 3
    flags.fillFromArgs(parser=parser)
    flags.lock()

    # Initialize configuration object, add accumulator, merge, and run.
    from AthenaConfiguration.MainServicesConfig import MainServicesCfg
    cfg = MainServicesCfg(flags)

    from ByteStreamCnvSvc.ByteStreamConfig import ByteStreamReadCfg
    tileTypeNames = [f'TileDigitsContainer/{args.digits}']
    cfg.merge( ByteStreamReadCfg(flags, type_names = tileTypeNames) )

    tileDigitsFlxMonitorAccumulator  = TileDigitsFlxMonitoringConfig(flags,
                                                                     fragIDs = fragIDs,
                                                                     TileDigitsContainerLegacy=args.digits,
                                                                     TileDigitsContainerFlx="TileDigitsCnt")

    cfg.merge(tileDigitsFlxMonitorAccumulator)

    # Any last things to do?
    if args.postExec:
        log.info('Executing postExec: %s', args.postExec)
        exec(args.postExec)

    cfg.printConfig(withDetails = True, summariseProps = True)
    flags.dump()

    cfg.store( open('TileDigitsFlxMonitorAlgorithm.pkl','wb') )

    sc = cfg.run()

    import sys
    # Success should be 0
    sys.exit(not sc.isSuccess())

File: TileCalorimeter/TileMonitoring/src/TileDigitsFlxMonitorAlgorithm.cxx
/*
  Copyright (C) 2002-2023 CERN for the benefit of the ATLAS collaboration
*/

// Tile includes
#include "TileDigitsFlxMonitorAlgorithm.h"
#include "TileIdentifier/TileHWID.h"
#include "TileCalibBlobObjs/TileCalibUtils.h"

// Athena includes
#include "StoreGate/ReadHandle.h"

#include <cmath>
#include <algorithm>


StatusCode TileDigitsFlxMonitorAlgorithm::initialize() {

  ATH_CHECK( AthMonitorAlgorithm::initialize() );

  ATH_MSG_INFO("in initialize()");

  ATH_CHECK( detStore()->retrieve(m_tileHWID) );
  ATH_CHECK( m_digitsContainerKeyLegacy.initialize() );
  ATH_CHECK( m_digitsContainerKeyFlx.initialize() );

  std::ostringstream os;

  if ( m_fragIDsToCompare.size() != 0) {
    std::sort(m_fragIDsToCompare.begin(), m_fragIDsToCompare.end());
    for (int fragID : m_fragIDsToCompare) {
      unsigned int ros    = fragID >> 8;
      unsigned int drawer = fragID & 0xFF;
      std::string module = TileCalibUtils::getDrawerString(ros, drawer);
      os << " " << module << "/0x" << std::hex << fragID << std::dec;
    }
  } else {
    os << "NONE";
  }
  m_firstFelix = static_cast<unsigned int>(m_firstSample) + m_felixOffset;
  m_nSamples = (m_lastSample >= m_firstSample) ? 1 + m_lastSample - m_firstSample : 0;

  ATH_MSG_INFO("Monitored modules/frag ID:" << os.str());

  return StatusCode::SUCCESS;
}


StatusCode TileDigitsFlxMonitorAlgorithm::fillHistograms( const EventContext& ctx ) const {

  // In case you want to measure the execution time
  auto timer = Monitored::Timer("TIME_execute");

  SG::ReadHandle<TileDigitsContainer> digitsContainerLegacy(m_digitsContainerKeyLegacy, ctx);
  ATH_CHECK( digitsContainerLegacy.isValid() );

  SG::ReadHandle<TileDigitsContainer> digitsContainerFlx(m_digitsContainerKeyFlx, ctx);
  ATH_CHECK( digitsContainerFlx.isValid() );

  std::array<std::string, 2> gainName{"_LG", "_HG"};

  std::vector<float>digitsLegacy[TileCalibUtils::MAX_CHAN][TileCalibUtils::MAX_GAIN];
  std::vector<float>digitsFlx[TileCalibUtils::MAX_CHAN][TileCalibUtils::MAX_GAIN];
  int found[TileCalibUtils::MAX_CHAN][TileCalibUtils::MAX_GAIN];

  const TileFragHash& hashFunc = digitsContainerFlx->hashFunc();

  for (int fragID : m_fragIDsToCompare) {

    memset(found,0,sizeof(found));

    IdentifierHash hash = static_cast<IdentifierHash>(hashFunc(fragID));
    unsigned int drawer = (fragID & 0x3F);
    unsigned int ros = fragID >> 8;
    std::string module = TileCalibUtils::getDrawerString(ros, drawer);
    std::array<std::string, 2> moduleName{module+gainName[0], module+gainName[1]};

    // Legacy digits
    const TileDigitsCollection* digitsCollectionLegacy = digitsContainerLegacy->indexFindPtr(hash);
    for (const TileDigits* tile_digits : *digitsCollectionLegacy) {

      ATH_MSG_VERBOSE((std::string) *tile_digits);

      unsigned int nSamples = tile_digits->nsamples();
      if (m_firstSample < nSamples) {

        HWIdentifier adcId = tile_digits->adc_HWID();
        int channel = m_tileHWID->channel(adcId);
        int gain = m_tileHWID->adc(adcId);

        std::string sampleName = module + "_ch_" + std::to_string(channel) + gainName[gain] +  "_samples";
        auto channelSample = Monitored::Scalar<float>(sampleName, 0.0F);

        std::vector<float> digits_monitored;
        auto it = tile_digits->samples().begin() + m_firstSample;
        auto end = tile_digits->samples().end();
        for(auto i = m_nSamples; it != end && i-->0; ++it) {
          float sample = (*it);
          channelSample = sample;
          fill("TileLegacyMonSamples", channelSample);
          digits_monitored.push_back(sample);
        }

        nSamples = digits_monitored.size();
        if (nSamples>0) {

          std::string channelName = moduleName[gain] + "_channel";
          auto monitoredChannel = Monitored::Scalar<float>(channelName, channel);

          std::string monPedestal = moduleName[gain] + "_Pedestal";
          auto pedestal = Monitored::Scalar<float>(monPedestal, digits_monitored[0]);
          fill("TileLegacyMonPed", monitoredChannel, pedestal);

          if (nSamples>1) {

            double sampleSum = std::accumulate( digits_monitored.begin(), digits_monitored.end(), 0.0);

            double sampleRMS = std::accumulate( digits_monitored.begin(), digits_monitored.end(), 0.0, [] (double acc, float sample) {
                  return acc + sample * sample;
                });
            sampleRMS -= sampleSum * sampleSum / nSamples;
            sampleRMS = (sampleRMS > 0.0) ? std::sqrt(sampleRMS / (nSamples - 1)) : 0.0;

            std::string monHFN = moduleName[gain] + "_HFN";
            auto hfn = Monitored::Scalar<float>(monHFN, sampleRMS);
            fill("TileLegacyMonHFN", monitoredChannel, hfn);
          }

          found[channel][gain] |= 1;
          digitsLegacy[channel][gain].swap(digits_monitored);
        }
      }
    }

    // FELIX digits
    const TileDigitsCollection* digitsCollectionFlx = digitsContainerFlx->indexFindPtr(hash);
    for (const TileDigits* tile_digits : *digitsCollectionFlx) {

      ATH_MSG_VERBOSE((std::string) *tile_digits);

      unsigned int nSamples = tile_digits->nsamples();
      if (m_firstFelix < nSamples) {

        HWIdentifier adcId = tile_digits->adc_HWID();
        int channel = m_tileHWID->channel(adcId);
        int gain = m_tileHWID->adc(adcId);

        std::string sampleName = module + "_ch_" + std::to_string(channel) + gainName[gain] +  "_samples";
        auto channelSample = Monitored::Scalar<float>(sampleName, 0.0F);

        std::vector<float> digits_monitored;
        auto it = tile_digits->samples().begin() + m_firstFelix;
        auto end = tile_digits->samples().end();
        
        for(auto i = m_nSamples; it != end && i-->0; ++it) {
          float sample = (*it);
          channelSample = sample;
          fill("TileFlxMonSamples", channelSample);
          digits_monitored.push_back(sample);
        }

        nSamples = digits_monitored.size();
        if (nSamples>0) {

          std::string channelName = moduleName[gain] + "_channel";
          auto monitoredChannel = Monitored::Scalar<float>(channelName, channel);

          std::string monPedestal = moduleName[gain] + "_Pedestal";
          auto pedestal = Monitored::Scalar<float>(monPedestal, digits_monitored[0]);
          fill("TileFlxMonPed", monitoredChannel, pedestal);

          if (nSamples>1) {

            double sampleSum = std::accumulate( digits_monitored.begin(), digits_monitored.end(), 0.0);

            double sampleRMS = std::accumulate( digits_monitored.begin(), digits_monitored.end(), 0.0, [] (double acc, float sample) {
                  return acc + sample * sample;
                });
            sampleRMS -= sampleSum * sampleSum / nSamples;
            sampleRMS = (sampleRMS > 0.0) ? std::sqrt(sampleRMS / (nSamples - 1)) : 0.0;

            std::string monHFN = moduleName[gain] + "_HFN";
            auto hfn = Monitored::Scalar<float>(monHFN, sampleRMS);
            fill("TileFlxMonHFN", monitoredChannel, hfn);
          }

          found[channel][gain] |= 2;
          digitsFlx[channel][gain].swap(digits_monitored);
        }
      }
    }

    // Compare digits from Legacy and FELIX
    for(unsigned int gain = 0; gain<TileCalibUtils::MAX_GAIN; ++gain) {

      std::string channelName      = moduleName[gain] + "_channel";
      std::string moduleSampleName = moduleName[gain] + "_samples_diff";

      auto monitoredChannel = Monitored::Scalar<float>(channelName, 0.0F);
      auto moduleSampleDiff = Monitored::Scalar<float>(moduleSampleName, 0.0F);

      for(unsigned int channel = 0; channel<TileCalibUtils::MAX_CHAN; ++channel) {

        if (found[channel][gain] == 3) {

          monitoredChannel = channel;

          std::string sampleName = module + "_ch_" + std::to_string(channel) + gainName[gain] +  "_samples_diff";
          auto channelSampleDiff = Monitored::Scalar<float>(sampleName, 0.0F);

          for (auto it1=digitsLegacy[channel][gain].begin(), it2=digitsFlx[channel][gain].begin();
               it1!=digitsLegacy[channel][gain].end() && it2!=digitsFlx[channel][gain].end(); ++it1, ++it2) {

            float diff = (*it2)- (*it1)*m_felixScale;
            channelSampleDiff = diff;
            moduleSampleDiff = diff;

            fill("TileChannelAllSamplesDiff", channelSampleDiff);
            fill("TileModuleAllSamplesDiff", moduleSampleDiff);
            fill("TileModuleSamplesDiff", monitoredChannel, moduleSampleDiff);
          }
        }
      }
    }
  }

  fill("TileDigitsFlxMonExecuteTime", timer);

  return StatusCode::SUCCESS;
}
